### PR TITLE
ISSUE-624 Apisix - SLO

### PR DESCRIPTION
**Pull request by vttranlina**
resolve https://github.com/linagora/james-project-private/issues/624

### CODE REVIEW

File: demo/apisix/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,4 @@
+FROM apache/apisix:3.2.0-debian
```
Comment (vttranlina) [inline demo/apisix/Dockerfile:1]: ## Why do we need it?

1) Apisix and Plugin Runner must be run in the same container
2) If we use `apache/apisix:3.2.0-debian`, we got an error when running. Ref (https://github.com/apache/apisix-java-plugin-runner/issues/165)
3) Docker image for this does not exist in apisix dockerhub

**Comment by vttranlina** [comment]
Test succeed

- `bash test.sh`

```
+ REDIRECT_URI=http://test.sso.example.com:8080/login-callback.html
+ curl http://sso.example.com/
+ grep token
+ perl -pe s/.*?value="(\w+)".*$/$1/
+ head -n1
+ TOKEN=1682598090_52514
+ curl -X POST -d user=james-user@tmail.com -d password=secret -d lmAuth=LDAP -d token=1682598090_52514 -H Accept: application/json http://sso.example.com/
+ perl -pe s/.*"id":"(.*?)".*$/$1/
+ COOKIE=9b7075dcd8c2fb1a59e19438ad2f273c683ce474a55edd205a134556feb3bea9
+ echo Identified by LLNG
Identified by LLNG
+ curl -s -D - -o /dev/null -b lemonldap=9b7075dcd8c2fb1a59e19438ad2f273c683ce474a55edd205a134556feb3bea9 http://sso.example.com/oauth2/authorize?response_type=code&client_id=james&scope=openid+profile+email&confirm=1&redirect_uri=http://test.sso.example.com:8080/login-callback.html
+ grep Location
+ perl -pe s/.*?code=(\w+).*?$/$1/
+ CODE=39a5dc274f692c19327299ff827970dbe5401208a27838aff0ae1cc292090313
+ echo Got a code in authorization flow (39a5dc274f692c19327299ff827970dbe5401208a27838aff0ae1cc292090313)
Got a code in authorization flow (39a5dc274f692c19327299ff827970dbe5401208a27838aff0ae1cc292090313)
+ curl -X POST -d grant_type=authorization_code -d redirect_uri=http://test.sso.example.com:8080/login-callback.html -d code=39a5dc274f692c19327299ff827970dbe5401208a27838aff0ae1cc292090313 -u james:james http://sso.example.com/oauth2/token
+ perl -pe s/^.*"access_token"\s*:\s*"(.*?)".*$/$1/
+ ACCESS_TOKEN=eyJhbGciOiJSUzUxMiIsImtpZCI6ImRyakcxdDU3NnpVVzZnckdxMVU0QUEiLCJ0eXAiOiJhdCtKV1QifQ.eyJzdWIiOiJqYW1lcy11c2VyQHRtYWlsLmNvbSIsImp0aSI6IjA0OTI3YTRmYTdlNjM2NDA1OWFhZjQwZTg1OTU5NjZmNzViNzNhZThlOWU4ZDBjNGE4NWU5YzU3YTM2ZjMyNTciLCJzaWQiOiJpNndYTlBzTlhNbG8xOWZCNTcrTEVHUjJPQSt6NlZGTUhBaGUrSDQ4bUVjIiwiZXhwIjoxNjgyNjczNTcwLCJlbWFpbCI6ImphbWVzLXVzZXJAdG1haWwuY29tIiwibmFtZSI6ImphbWVzLXVzZXIiLCJhdWQiOlsiamFtZXMiXSwic2NvcGUiOiJvcGVuaWQgcHJvZmlsZSBlbWFpbCIsImlhdCI6MTY4MjY2OTk3MCwiaXNzIjoiaHR0cDovL3Nzby5leGFtcGxlLmNvbSIsImNsaWVudF9pZCI6ImphbWVzIn0.NB7lCRNCalshjz72HCqugP8yir54rAuo44VpSu9GO-pDFt2V61711-SCdZ6Znfa9oQL7VMaaCTByrNbS4CksfmA7KovRS0Ig9Zhc2ZdZEPlIEeKWUVicjAbeOSTNMlF3iB0a9OvMeJh1e1B_n0nFvqizti1UXXXO-t6VWwEyQsFTd-2PZcOvElvS3MvRYzR3Ex_qWgDJ9RROmBk9K7asvzdwpLm4ld69riKxBzM_yHUIDILpWXLeyvfR04qI7K1ymKHjdXotOUf1wCKvPA3mdsxUnNdViRH6jyrvdql055qsbD2HFb79FErMgucw5StEdSpdThjvUdUArSQ1lKWflQ
+ echo Got an access_token
Got an access_token
+ curl -H Authorization: Bearer eyJhbGciOiJSUzUxMiIsImtpZCI6ImRyakcxdDU3NnpVVzZnckdxMVU0QUEiLCJ0eXAiOiJhdCtKV1QifQ.eyJzdWIiOiJqYW1lcy11c2VyQHRtYWlsLmNvbSIsImp0aSI6IjA0OTI3YTRmYTdlNjM2NDA1OWFhZjQwZTg1OTU5NjZmNzViNzNhZThlOWU4ZDBjNGE4NWU5YzU3YTM2ZjMyNTciLCJzaWQiOiJpNndYTlBzTlhNbG8xOWZCNTcrTEVHUjJPQSt6NlZGTUhBaGUrSDQ4bUVjIiwiZXhwIjoxNjgyNjczNTcwLCJlbWFpbCI6ImphbWVzLXVzZXJAdG1haWwuY29tIiwibmFtZSI6ImphbWVzLXVzZXIiLCJhdWQiOlsiamFtZXMiXSwic2NvcGUiOiJvcGVuaWQgcHJvZmlsZSBlbWFpbCIsImlhdCI6MTY4MjY2OTk3MCwiaXNzIjoiaHR0cDovL3Nzby5leGFtcGxlLmNvbSIsImNsaWVudF9pZCI6ImphbWVzIn0.NB7lCRNCalshjz72HCqugP8yir54rAuo44VpSu9GO-pDFt2V61711-SCdZ6Znfa9oQL7VMaaCTByrNbS4CksfmA7KovRS0Ig9Zhc2ZdZEPlIEeKWUVicjAbeOSTNMlF3iB0a9OvMeJh1e1B_n0nFvqizti1UXXXO-t6VWwEyQsFTd-2PZcOvElvS3MvRYzR3Ex_qWgDJ9RROmBk9K7asvzdwpLm4ld69riKxBzM_yHUIDILpWXLeyvfR04qI7K1ymKHjdXotOUf1wCKvPA3mdsxUnNdViRH6jyrvdql055qsbD2HFb79FErMgucw5StEdSpdThjvUdUArSQ1lKWflQ http://sso.example.com/oauth2/userinfo
+ grep james
+ echo Access_token is valid
Access_token is valid
+ echo -n Trying James: 
Trying James: + APISIX_JMAP_ENDPOINT=apisix.example.com:9080/oidc/jmap/session
+ curl -v -H Accept: application/json; jmapVersion=rfc-8621 -H Authorization: Bearer eyJhbGciOiJSUzUxMiIsImtpZCI6ImRyakcxdDU3NnpVVzZnckdxMVU0QUEiLCJ0eXAiOiJhdCtKV1QifQ.eyJzdWIiOiJqYW1lcy11c2VyQHRtYWlsLmNvbSIsImp0aSI6IjA0OTI3YTRmYTdlNjM2NDA1OWFhZjQwZTg1OTU5NjZmNzViNzNhZThlOWU4ZDBjNGE4NWU5YzU3YTM2ZjMyNTciLCJzaWQiOiJpNndYTlBzTlhNbG8xOWZCNTcrTEVHUjJPQSt6NlZGTUhBaGUrSDQ4bUVjIiwiZXhwIjoxNjgyNjczNTcwLCJlbWFpbCI6ImphbWVzLXVzZXJAdG1haWwuY29tIiwibmFtZSI6ImphbWVzLXVzZXIiLCJhdWQiOlsiamFtZXMiXSwic2NvcGUiOiJvcGVuaWQgcHJvZmlsZSBlbWFpbCIsImlhdCI6MTY4MjY2OTk3MCwiaXNzIjoiaHR0cDovL3Nzby5leGFtcGxlLmNvbSIsImNsaWVudF9pZCI6ImphbWVzIn0.NB7lCRNCalshjz72HCqugP8yir54rAuo44VpSu9GO-pDFt2V61711-SCdZ6Znfa9oQL7VMaaCTByrNbS4CksfmA7KovRS0Ig9Zhc2ZdZEPlIEeKWUVicjAbeOSTNMlF3iB0a9OvMeJh1e1B_n0nFvqizti1UXXXO-t6VWwEyQsFTd-2PZcOvElvS3MvRYzR3Ex_qWgDJ9RROmBk9K7asvzdwpLm4ld69riKxBzM_yHUIDILpWXLeyvfR04qI7K1ymKHjdXotOUf1wCKvPA3mdsxUnNdViRH6jyrvdql055qsbD2HFb79FErMgucw5StEdSpdThjvUdUArSQ1lKWflQ apisix.example.com:9080/oidc/jmap/session
+ grep uploadUrl
+ echo OK
OK
+ Logout

```

File: demo/docker-compose.yml [hunk 2]
```diff
@@ -104,14 +104,16 @@ services:
 
   apisix:
     container_name: apisix.example.com
-    image: apache/apisix:3.2.0-debian
+    image: linagora/apisix:3.2.0-debian-javaplugin
```
Comment (quantranhong1999) [inline demo/docker-compose.yml:107]: IMO we should plan to let the CI build and publish the image

**Review by quantranhong1999 — COMMENTED** [review]
Great job so far <3

Can we have a UI demo for this?

**Comment by Arsnael** [comment]
A pity we gonna need to maintain an apisix image for adding plugins but I guess it's ok in this case.

Good job @vttranlina , looking good :)

**Comment by Arsnael** [comment]
Side note: it's good I did not advance too much on the helm topic, this means now we clearly need our own Helm chart for running this :)

**Comment by psouquet** [comment]
Hi, took a look at this PR as cross team review and saw nothing wrong here, good work :)

**Comment by vttranlina** [comment]
rebase master & implement redis for plugin

File: demo/apisix/conf/config.yaml [hunk 3]
```diff
@@ -5,4 +5,4 @@ deployment:
 
 ext-plugin:
   path_for_test: /tmp/runner.sock
-  cmd: ['java', '-jar', '-Xmx1g', '-Xms1g', '/usr/local/apisix/token_revoker_plugin.jar']
\ No newline at end of file
+  cmd: ['java', '-jar', '-DREDIS_HOST=redis.example.com', '-DREDIS_PORT=6379', '-Xmx1g', '-Xms1g', '/usr/local/apisix/token_revoker_plugin.jar']
```
Comment (quantranhong1999) [inline demo/apisix/conf/config.yaml:8]: How about the Redis authentication?
E.g Redis password was needed on our preprod:
```
redisURL=redis://${env:OP_REDIS_PASSWORD}@${env:OP_REDIS_URL}:${env:OP_REDIS_PORT}
```

File: demo/apisix/tmail-apisix-plugin-runner/src/test/java/com/linagora/apisix/plugin/RevokedTokenRepositoryContract.java [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+
+    @Test
+    default void existShouldReturnTrueWhenExist() {
+        String sid = "sid1";
+        testee().add(sid);
+        assertThat(testee().exist(sid)).isTrue();
+    }
+
+    @Test
+    default void existShouldBeIdempotent() {
```
Comment (Arsnael) [inline demo/apisix/tmail-apisix-plugin-runner/src/test/java/com/linagora/apisix/plugin/RevokedTokenRepositoryContract.java:25]: such test does not make sense, it's a get after all on something in redis. I would expect more a test about addShouldBeIdempotent ?

**Review by Arsnael — COMMENTED** [review]
Good job otherwise :)

File: demo/apisix/Dockerfile [hunk 5]
```diff
@@ -0,0 +1,4 @@
...
+USER root
+RUN apt -y install openjdk-11-jdk
```
Comment (chibenwa) [inline demo/apisix/Dockerfile:3]: JDK is needed or JRE is enough?

File: demo/apisix/tmail-apisix-plugin-runner/src/main/java/com/linagora/apisix/plugin/TokenRevokedFilter.java [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+        logger.debug("Received a new request");
+        sidFilter(request, response);
+        chain.filter(request, response);
+    }
+
+    private void sidFilter(HttpRequest request, HttpResponse response) {
+        Optional.ofNullable(request.getHeader("Authorization"))
+            .or(() -> Optional.ofNullable(request.getHeader("authorization")))
+            .map(String::trim)
+            .map(bearerToken -> bearerToken.substring(7)) // remove `Bearer `
```
Comment (chibenwa) [inline demo/apisix/tmail-apisix-plugin-runner/src/main/java/com/linagora/apisix/plugin/TokenRevokedFilter.java:42]: Assert that the 7 first char matches `Bearer ` before removing them...

File: demo/docker-compose.yml (hunk 2 above)
Comment (chibenwa) [reply demo/docker-compose.yml:107]: Can we remove `jwt-revoker` container?

Comment (Arsnael) [reply demo/docker-compose.yml:107]: Logically krakend too then :)

**Comment by vttranlina** [comment]
https://user-images.githubusercontent.com/81145350/236380870-f84b3809-7093-4219-adb7-639042e35935.mp4

UI Test

File: demo/apisix/conf/apisix.yaml [hunk 7]
```diff
@@ -162,4 +158,8 @@ plugin_configs:
           failures: 3  # should be var: JMAP_CIRCUIT_BREAKER_MAXERRORS
         healthy:
           successes: 1
+      ext-plugin-pre-req:
+        conf:
+          - name: TokenRevokedFilter
+            value: '{"enable":"feature"}'
```
Comment (quantranhong1999) [inline demo/apisix/conf/apisix.yaml:165]: Declaring the token revoked plugin in `jmap-plugin` would enforce the basic routes to revoke the jwt token as well (which is not needed).
Maybe this one should be explicitly declared?

File: demo/apisix/tmail-apisix-plugin-runner/src/main/java/com/linagora/apisix/plugin/AppConfiguration.java [hunk 8]
```diff
@@ -16,11 +18,12 @@
 public class AppConfiguration {
     private final Logger logger = LoggerFactory.getLogger(AppConfiguration.class);
 
-    public static StringRedisTemplate redisTemplate(String host, int port) {
+    public static StringRedisTemplate redisTemplate(String host, int port, Optional<String> password) {
         var redisConfiguration = new RedisStandaloneConfiguration();
```
Comment (quantranhong1999) [inline demo/apisix/tmail-apisix-plugin-runner/src/main/java/com/linagora/apisix/plugin/AppConfiguration.java:22]: Can this `RedisStandaloneConfiguration` work with the Master/Slave topology we are using?

Comment (vttranlina) [reply demo/apisix/tmail-apisix-plugin-runner/src/main/java/com/linagora/apisix/plugin/AppConfiguration.java:22]: NO?
"Standalone" here is client-side. 
And...
image

File: demo/apisix/conf/apisix.yaml (hunk 7 above)
Comment (vttranlina) [reply demo/apisix/conf/apisix.yaml:165]: I got your meaning, 
Now I'm confusing: 
- Each route has only one `plugin_config_id`, if we declare  `ext-plugin-pre-req` for each route, it is quite verbose
- The `TokenRevokedFilter` will filter requests has authorization header start with `Bearer`, is it fine? IMO it is no problem

Comment (quantranhong1999) [reply demo/apisix/conf/apisix.yaml:165]: Rethinking the basic routes is basically for testing purposes. I think it is fine too. As long as this does not break the basic routes flow.

Comment (Arsnael) [reply demo/apisix/conf/apisix.yaml:165]: Did you try to add an other `plugin_config_id` field in the routes? I didn't test but I would not be surprised it works :)

File: demo/apisix/tmail-apisix-plugin-runner/src/main/java/com/linagora/apisix/plugin/AppConfiguration.java (hunk 8 above)
Comment (Arsnael) [reply demo/apisix/tmail-apisix-plugin-runner/src/main/java/com/linagora/apisix/plugin/AppConfiguration.java:22]: @vttranlina I think @quantranhong1999 makes sense.

I don't say remove the standalone, it's perfect for the demo for example, but we need to manage master/slave mode as well, if we want to use it on preprod/prod

File: demo/apisix/conf/apisix.yaml (hunk 7 above)
Comment (vttranlina) [reply demo/apisix/conf/apisix.yaml:165]: I tried to with below syntax, and it does not work
```yaml
    plugin_config_id: 
      - jmap-plugin
      - other-plugin
```

Comment (Arsnael) [reply demo/apisix/conf/apisix.yaml:165]: What about this:

```
plugin_config_id: jmap-plugin
plugin_config_id: other-plugin
```
? As I see we can put `plugins` and add a `plugin_config_id` that is defining something similar, maybe can put more than one field :) Maybe not :)

File: demo/apisix/tmail-apisix-plugin-runner/src/main/java/com/linagora/apisix/plugin/AppConfiguration.java (hunk 8 above)
Comment (vttranlina) [reply demo/apisix/tmail-apisix-plugin-runner/src/main/java/com/linagora/apisix/plugin/AppConfiguration.java:22]: > "Standalone" here is client-side.

We have only one endpoint of Redis on preprod/prod for james, right? 
I think we must be care of another topo when we have >=2 endpoint

File: demo/apisix/conf/apisix.yaml (hunk 7 above)
Comment (vttranlina) [reply demo/apisix/conf/apisix.yaml:165]: NO, it is not a valid yaml syntax

File: demo/apisix/tmail-apisix-plugin-runner/src/main/java/com/linagora/apisix/plugin/AppConfiguration.java (hunk 8 above)
Comment (Arsnael) [reply demo/apisix/tmail-apisix-plugin-runner/src/main/java/com/linagora/apisix/plugin/AppConfiguration.java:22]: I remember it being not that simple... Pretty sure we thought like this first for Redis in James as well and we had to do the differentiation at the end :)

Comment (Arsnael) [reply demo/apisix/tmail-apisix-plugin-runner/src/main/java/com/linagora/apisix/plugin/AppConfiguration.java:22]: Actually you write only on master but should be able to read from the slaves I believe?

Comment (vttranlina) [reply demo/apisix/tmail-apisix-plugin-runner/src/main/java/com/linagora/apisix/plugin/AppConfiguration.java:22]: let me check the similarities in James' code

File: demo/apisix/tmail-apisix-plugin-runner/src/main/java/com/linagora/apisix/plugin/RedisRevokedTokenRepository.java [hunk 9]
```diff
@@ -0,0 +1,26 @@
+package com.linagora.apisix.plugin;
+
+import java.time.Duration;
+
+import org.springframework.data.redis.core.StringRedisTemplate;
+
+public class RedisRevokedTokenRepository implements IRevokedTokenRepository {
+    public static final Duration CACHE_DURATION = Duration.ofMinutes(60);
```
Comment (chibenwa) [inline demo/apisix/tmail-apisix-plugin-runner/src/main/java/com/linagora/apisix/plugin/RedisRevokedTokenRepository.java:8]: This duration is likely too short and should exceed refresh token validity to NOT expose us to security weaknesses.


I propose 1 week as a minimum...

File: demo/apisix/tmail-apisix-plugin-runner/src/main/java/com/linagora/apisix/plugin/TokenRevokedFilter.java [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+                    logger.debug("Token has been revoked, Sid: " + sid);
+                    makeUnAuthorizedRequest(request, response);
+                } else {
+                    logger.debug("Token valid, Sid: " + sid);
+                }
+            });
+    }
+
+    private void makeUnAuthorizedRequest(HttpRequest request, HttpResponse response) {
+        // TODO refactor
```
Comment (chibenwa) [inline demo/apisix/tmail-apisix-plugin-runner/src/main/java/com/linagora/apisix/plugin/TokenRevokedFilter.java:56]: ?

File: demo/apisix/tmail-apisix-plugin-runner/src/test/java/com/linagora/apisix/plugin/ClusterRedisRevokedTokenRepositoryTest.java [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+        REDIS_REPLICA.stop();
+    }
+
+    RedisRevokedTokenRepository revokedTokenRepository;
+
+    @BeforeEach
+    void setup() {
+        revokedTokenRepository = new RedisRevokedTokenRepository(AppConfiguration.initRedisCommand(
+            String.format("redis://%s@localhost:%d",
+                "my_master_password", REDIS_MASTER.getMappedPort(6379)), false));
```
Comment (vttranlina) [inline demo/apisix/tmail-apisix-plugin-runner/src/test/java/com/linagora/apisix/plugin/ClusterRedisRevokedTokenRepositoryTest.java:59]: This testes success now. 
But fail if the input is list uri (master & slave). 

I'm stuck here, I guess maybe by Redis configure not correct somewhere

**Review by chibenwa — COMMENTED** [review]
Please add a README on how to enable the APISIX plugin, and how to connect to a Redis cluster...

File: demo/apisix/tmail-apisix-plugin-runner/src/main/java/com/linagora/apisix/plugin/AppConfiguration.java [hunk 12]
```diff
@@ -0,0 +1,60 @@
...
+        logger.info("The plugin using local memory for storage revoked tokens");
+        return new IRevokedTokenRepository.MemoryRevokedTokenRepository();
+    }
+
+    public static RedisStringCommands<String, String> initRedisCommand(String redisUrl, boolean redisClusterEnable) {
+        List<RedisURI> redisURIList = Arrays.stream(redisUrl.split(","))
+            .map(RedisURI::create)
+            .collect(Collectors.toList());
+        if (!redisClusterEnable) {
+            return RedisClient.create(redisURIList.get(0)).connect().sync();
```
Comment (chibenwa) [inline demo/apisix/tmail-apisix-plugin-runner/src/main/java/com/linagora/apisix/plugin/AppConfiguration.java:52]: We should fail if several hosts are specified while we do not support the cluster mode...

File: demo/docker-compose.yml [hunk 13]
```diff
@@ -80,38 +80,28 @@ services:
...
       - ./apisix/conf/config.yaml:/usr/local/apisix/conf/config.yaml
+      - ./apisix/tmail-apisix-plugin-runner/target/tmail-apisix-plugin-runner-0.0.1-SNAPSHOT.jar:/usr/local/apisix/token_revoker_plugin.jar
     networks:
       - tmail
     ports:
       - "9080:9080/tcp"
+      - "8081:8081"
+
+  redis:
+    container_name: redis.example.com
```
Comment (quantranhong1999) [inline demo/docker-compose.yml:97]: Add this into /etc/hosts instruction?

**Review by chibenwa — APPROVED** [review]

File: demo/docker-compose.yml (hunk 13 above)
Comment (vttranlina) [reply demo/docker-compose.yml:97]: not need, backend uses it, end-user doesn't need to know it.

**Review by quantranhong1999 — APPROVED** [review]

**Review by Arsnael — APPROVED** [review]

**Comment by vttranlina** [comment]
rebase & squash fixup

**Pull request merged**